File: server/dotnet/sdk-dotnet-example/Presentation/Validators/Rules/IbanRules.cs
using System.Text.RegularExpressions;
using Business.Domain.Common.Enums;
using FluentValidation;

namespace Presentation.Validators.Rules;

internal static class IbanRules
{
    private static readonly Regex BasicIbanRx =
        new(@"^[A-Z]{2}\d{2}[A-Z0-9]+$", RegexOptions.Compiled);
    
    public static string Clean(string? iban) =>
        new string((iban ?? string.Empty).Where(char.IsLetterOrDigit).ToArray()).ToUpperInvariant();
    
    private static (string Prefix, int Length)? SpecFor(Country country) =>
        country switch
        {
            Country.France  => ("FR", 27),
            Country.Germany => ("DE", 22),
            Country.England => ("GB", 22),
            _               => null
        };

    private static bool HasValidChecksum(string iban)
    {
        var rear = iban[4..] + iban[..4];

        int remainder = 0;
        foreach (var ch in rear)
        {
            var digits = char.IsLetter(ch) ? ((ch - 'A') + 10).ToString() : ch.ToString();
            foreach (var d in digits)
            {
                remainder = (remainder * 10 + (d - '0')) % 97;
            }
        }
        return remainder == 1;
    }

    public static IRuleBuilderOptions<T, string?> MustHaveBasicIbanFormat<T>(
        this IRuleBuilder<T, string?> rule)
    {
        return rule
            .Must(value =>
            {
                if (string.IsNullOrWhiteSpace(value)) return true;
                var iban = Clean(value);
                return BasicIbanRx.IsMatch(iban);
            })
            .WithMessage("IBAN format is invalid (expected: 2 letters country + 2 digits + alphanumerics).");
    }

    public static IRuleBuilderOptions<T, string?> MustMatchIbanCountryAndLength<T>(
        this IRuleBuilder<T, string?> rule,
        Func<T, Country> countrySelector)
    {
        return rule
            .Must((obj, value) =>
            {
                if (string.IsNullOrWhiteSpace(value)) return true; 
                var spec = SpecFor(countrySelector(obj));
                if (spec is null) return false;

                var iban = Clean(value);
                return iban.StartsWith(spec.Value.Prefix) && iban.Length == spec.Value.Length;
            })
            .WithMessage(obj =>
            {
                var c = countrySelector(obj);
                var spec = SpecFor(c);
                return spec is null
                    ? "IBAN country is not supported."
                    : $"IBAN must start with '{spec.Value.Prefix}' and be {spec.Value.Length} characters for {c}.";
            });
    }

    public static IRuleBuilderOptions<T, string?> MustHaveValidIbanChecksum<T>(
        this IRuleBuilder<T, string?> rule)
    {
        return rule
            .Must(value =>
            {
                if (string.IsNullOrWhiteSpace(value)) return true;
                var iban = Clean(value);
                if (!BasicIbanRx.IsMatch(iban)) return false;
                return HasValidChecksum(iban);
            })
            .WithMessage("IBAN checksum is invalid.");
    }
}

File: server/nodejs/src/presentation/validators/rules/iban-rules.ts
import { Country } from '../../../business/domain/common/enums/country';

export class IbanRules {
    private static readonly BASIC_IBAN_REGEX = /^[A-Z]{2}\d{2}[A-Z0-9]+$/;

    private static readonly IBAN_LENGTHS: Record<string, number> = {
        FR: 27,
        DE: 22,
        GB: 22,
    };

    static clean(iban: string | null | undefined): string {
        if (iban === null || iban === undefined) {
            return '';
        }

        return iban.replace(/[^a-zA-Z0-9]/g, '').toUpperCase();
    }

    private static getCountrySpec(country?: Country | null): { prefix: string; length: number } | null {
        if (!country) {
            return null;
        }

        switch (country) {
            case Country.France:
                return { prefix: 'FR', length: this.IBAN_LENGTHS.FR };
            case Country.Germany:
                return { prefix: 'DE', length: this.IBAN_LENGTHS.DE };
            case Country.England:
                return { prefix: 'GB', length: this.IBAN_LENGTHS.GB };
            default:
                return null;
        }
    }

    private static hasValidChecksum(iban: string): boolean {
        const rear = iban.substring(4) + iban.substring(0, 4);

        let remainder = 0;
        for (const ch of rear) {
            const digits = /[A-Z]/.test(ch)
                ? String(ch.charCodeAt(0) - 'A'.charCodeAt(0) + 10)
                : ch;

            for (const d of digits) {
                remainder = (remainder * 10 + (d.charCodeAt(0) - '0'.charCodeAt(0))) % 97;
            }
        }

        return remainder === 1;
    }

    static validateIban(iban?: string | null, country?: Country | null): string[] {
        const errors: string[] = [];

        if (!iban || iban.trim() === '') {
            return errors;
        }

        const cleanedIban = this.clean(iban);

        if (!this.BASIC_IBAN_REGEX.test(cleanedIban)) {
            errors.push('IBAN format is invalid (expected: 2 letters country + 2 digits + alphanumerics).');
            return errors;
        }

        if (country !== null && country !== undefined) {
            const spec = this.getCountrySpec(country);

            if (!spec) {
                errors.push('IBAN country is not supported.');
            } else if (
                !cleanedIban.startsWith(spec.prefix) ||
                cleanedIban.length !== spec.length
            ) {
                errors.push(
                    `IBAN must start with '${spec.prefix}' and be ${spec.length} characters for ${country}.`
                );
            }
        }

        if (!this.hasValidChecksum(cleanedIban)) {
            errors.push('IBAN checksum is invalid.');
        }

        return errors;
    }
}